### PR TITLE
Adding PRF notebook to TESS folder

### PR DESCRIPTION
**Pull request by mustaric**
I've created a notebook to retrieve and examine the TESS pixel response function. 
Thanks,
Susan

### CODE REVIEW

File: notebooks/MAST/TESS/interm_tess_prf_retrieve/tess_prf_retrieval.ipynb [hunk 1]
```diff
@@ -0,0 +1,852 @@
...
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def pathLookup(ccd, camera, sector):\n",
+    "    \"\"\"\n",
+    "    Gets the datestring and the subdirectory for the specified PRF. \n",
+    "    The datestring and directory name can be found from the ccd, camera and sector.\n",
+    "    \n",
```
Comment (ceb8) [inline notebooks/MAST/TESS/interm_tess_prf_retrieve/tess_prf_retrieval.ipynb]: Add inputs and returns sections to docstring.

File: notebooks/MAST/TESS/interm_tess_prf_retrieve/tess_prf_retrieval.ipynb [hunk 2]
```diff
@@ -0,0 +1,852 @@
...
+    "    prfArray\n",
+    "        (np array) Full 117 x 117 interleaved prf Array for the requested file.\n",
+    "        \n",
+    "    \"\"\"\n",
+    "        \n",
+    "    fn = \"cam%u_ccd%u/tess%13s-prf-%1u-%1u-row%04u-col%04u.fits\" % \\\n",
+    "        (camera, ccd, datestring, camera, ccd, row, col)\n",
+    "\n",
+    "    filepath = os.path.join(path, fn)\n",
+    "    #print(filepath)\n",
```
Comment (ceb8) [inline notebooks/MAST/TESS/interm_tess_prf_retrieve/tess_prf_retrieval.ipynb]: Remove commented out line.

File: notebooks/MAST/TESS/interm_tess_prf_retrieve/tess_prf_retrieval.ipynb [hunk 3]
```diff
@@ -0,0 +1,852 @@
...
+    "    \"\"\"\n",
+    "\n",
+    "    posRows = np.array([1, 513, 1025, 1536, 2048])\n",
+    "    posCols = np.array([45, 557, 1069, 1580,2092])\n",
+    "\n",
+    "    difcol = np.abs(posCols - col)\n",
+    "    difrow = np.abs(posRows - row)\n",
+    "\n",
+    "    #Expand out to the four image position to interpolate between,\n",
+    "    #Return as a list of tuples.\n",
```
Comment (ceb8) [inline notebooks/MAST/TESS/interm_tess_prf_retrieve/tess_prf_retrieval.ipynb]: #Expand -> # Expand
#Return -> # Return

File: notebooks/MAST/TESS/interm_tess_prf_retrieve/tess_prf_retrieval.ipynb [hunk 4]
```diff
@@ -0,0 +1,852 @@
...
+    "    hdulistObj = fits.open(filepath)\n",
+    "    prfArray = hdulistObj[0].data\n",
+    "\n",
+    "    return prfArray\n",
+    "\n",
+    "def determineFourClosestPrfLoc(col, row):\n",
+    "    \"\"\"\n",
+    "    Determine the four pairs of col,row positions of your target. \n",
+    "    These are specific to TESS and where they chose to report their PRFs.\n",
+    "    Returns a list of (col,row) pairs.\n",
```
Comment (ceb8) [inline notebooks/MAST/TESS/interm_tess_prf_retrieve/tess_prf_retrieval.ipynb]: Add inputs and returns sections to docstring.

File: notebooks/MAST/TESS/interm_tess_prf_retrieve/tess_prf_retrieval.ipynb [hunk 5]
```diff
@@ -0,0 +1,852 @@
...
+   "outputs": [],
+   "source": [
+    "def getOffsetsFromPixelFractions(col, row):\n",
+    "    \"\"\"\n",
+    "    Determine just the fractional part (the intra-pixel part) of the col,row position.  \n",
+    "    For example, if (col, row) = (123.4, 987.6), then\n",
+    "    (colFrac, rowFrac) = (.4, .6). \n",
+    "    \n",
+    "    Returns the offset necessary for addressing the interleaved PRF array.\n",
+    "    to ensure you get the location appropriate for your sub-pixel values.\n",
```
Comment (ceb8) [inline notebooks/MAST/TESS/interm_tess_prf_retrieve/tess_prf_retrieval.ipynb]: Add inputs and returns sections to docstring.

File: notebooks/MAST/TESS/interm_tess_prf_retrieve/tess_prf_retrieval.ipynb [hunk 6]
```diff
@@ -0,0 +1,852 @@
...
+    "    The 13x13 pixel PRFs on at each grid location are sampled at a 9x9 intra-pixel grid, to\n",
+    "    describe how the PRF changes as the star moves by a fraction of a pixel in row or column.\n",
+    "    To extract out a single PRF, you need to address the 117x117 array in a funny way\n",
+    "    (117 = 13x9). Essentially you need to pull out every 9th element in the array, i.e.\n",
+    "\n",
+    "    .. code-block:: python\n",
+    "\n",
+    "        img = array[ [colOffset, colOffset+9, colOffset+18, ...],\n",
+    "                     [rowOffset, rowOffset+9, ...] ]\n",
+    "    \"\"\"\n",
```
Comment (ceb8) [inline notebooks/MAST/TESS/interm_tess_prf_retrieve/tess_prf_retrieval.ipynb]: Add inputs and returns sections to docstring.

File: notebooks/MAST/TESS/interm_tess_prf_retrieve/tess_prf_retrieval.ipynb [hunk 7]
```diff
@@ -0,0 +1,852 @@
...
+    "    gridSize = 9\n",
+    "\n",
+    "    assert colOffset < gridSize\n",
+    "    assert rowOffset < gridSize\n",
+    "\n",
+    "    #Number of pixels in regularly sampled PRF. Should be 13x13\n",
```
Comment (ceb8) [inline notebooks/MAST/TESS/interm_tess_prf_retrieve/tess_prf_retrieval.ipynb]: #Number -> # Number

File: notebooks/MAST/TESS/interm_tess_prf_retrieve/tess_prf_retrieval.ipynb [hunk 8]
```diff
@@ -0,0 +1,852 @@
...
+    "    r0 = imagePos[0][1]\n",
+    "    r1 = imagePos[2][1]\n",
+    "\n",
+    "    assert c0 != c1\n",
+    "    assert r0 != r1\n",
+    "\n",
+    "    dCol = (col-c0) / (c1-c0)\n",
+    "    dRow = (row-r0) / (r1 - r0)\n",
+    "\n",
+    "    #Intpolate across the rows\n",
```
Comment (ceb8) [inline notebooks/MAST/TESS/interm_tess_prf_retrieve/tess_prf_retrieval.ipynb]: #Intpolate -> # Interpolate

File: notebooks/MAST/TESS/interm_tess_prf_retrieve/tess_prf_retrieval.ipynb [hunk 9]
```diff
@@ -0,0 +1,852 @@
...
+    "    tmp1 = p11 + (p21 - p11) * dCol\n",
+    "    tmp2 = p12 + (p22 - p12) * dCol\n",
+    "\n",
+    "    #Interpolate across the columns\n",
```
Comment (ceb8) [inline notebooks/MAST/TESS/interm_tess_prf_retrieve/tess_prf_retrieval.ipynb]: #Interpolate -> # Interpolate

File: notebooks/MAST/TESS/interm_tess_prf_retrieve/tess_prf_retrieval.ipynb [hunk 10]
```diff
@@ -0,0 +1,852 @@
...
+    "\n",
+    "    Returns\n",
+    "    ---------\n",
+    "    A 13x13 numpy image array of the nearest PRF to the specifed column and row.\n",
+    "    \"\"\"\n",
+    "    col = float(col)\n",
+    "    row = float(row)\n",
+    "    prfImages = []\n",
+    "    \n",
+    "    #Determine a datestring in the file name and the path based on ccd/camer/sector\n",
```
Comment (ceb8) [inline notebooks/MAST/TESS/interm_tess_prf_retrieve/tess_prf_retrieval.ipynb]: #Determine -> # Determine

File: notebooks/MAST/TESS/interm_tess_prf_retrieve/tess_prf_retrieval.ipynb [hunk 11]
```diff
@@ -0,0 +1,852 @@
...
+    "    datestring, addPath = pathLookup(ccd, camera, sector)\n",
+    "    path = path + addPath\n",
+    "    \n",
+    "    #Convert the fractional pixels to the offset required for the interleaved pixels.\n",
```
Comment (ceb8) [inline notebooks/MAST/TESS/interm_tess_prf_retrieve/tess_prf_retrieval.ipynb]: #Convert -> # Convert

File: notebooks/MAST/TESS/interm_tess_prf_retrieve/tess_prf_retrieval.ipynb [hunk 12]
```diff
@@ -0,0 +1,852 @@
...
+    "    colOffset, rowOffset = getOffsetsFromPixelFractions(col, row)\n",
+    "    \n",
+    "    #Determine the 4 (col,row) locations with exact PRF measurements.\n",
```
Comment (ceb8) [inline notebooks/MAST/TESS/interm_tess_prf_retrieve/tess_prf_retrieval.ipynb]: #Determine -> # Determine

File: notebooks/MAST/TESS/interm_tess_prf_retrieve/tess_prf_retrieval.ipynb [hunk 13]
```diff
@@ -0,0 +1,852 @@
...
+    "\n",
+    "    Returns\n",
+    "    ---------\n",
+    "    A 13x13 numpy image array of the interpolated PRF.\n",
+    "    \"\"\"\n",
+    "    col = float(col)\n",
+    "    row = float(row)\n",
+    "    prfImages = []\n",
+    "    \n",
+    "    #Determine a datestring in the file name and the path based on ccd/camera/sector\n",
```
Comment (ceb8) [inline notebooks/MAST/TESS/interm_tess_prf_retrieve/tess_prf_retrieval.ipynb]: #Determine -> # Determine

File: notebooks/MAST/TESS/interm_tess_prf_retrieve/tess_prf_retrieval.ipynb [hunk 14]
```diff
@@ -0,0 +1,852 @@
...
+    "    datestring, subDirectory = pathLookup(ccd, camera, sector)\n",
+    "    path = path + subDirectory\n",
+    "    \n",
+    "    #Convert the fractional pixels to the offset required for the interleaved pixels.\n",
+    "    colOffset, rowOffset = getOffsetsFromPixelFractions(col, row)\n",
+    "    \n",
+    "    #Determine the 4 (col,row) locations with exact PRF measurements.\n",
```
Comment (ceb8) [inline notebooks/MAST/TESS/interm_tess_prf_retrieve/tess_prf_retrieval.ipynb]: #Determine -> # Determine

File: notebooks/MAST/TESS/interm_tess_prf_retrieve/tess_prf_retrieval.ipynb [hunk 15]
```diff
@@ -0,0 +1,852 @@
...
+    "    \"\"\"\n",
+    "    col = float(col)\n",
+    "    row = float(row)\n",
+    "    prfImages = []\n",
+    "    \n",
+    "    #Determine a datestring in the file name and the path based on ccd/camera/sector\n",
+    "    datestring, subDirectory = pathLookup(ccd, camera, sector)\n",
+    "    path = path + subDirectory\n",
+    "    \n",
+    "    #Convert the fractional pixels to the offset required for the interleaved pixels.\n",
```
Comment (ceb8) [inline notebooks/MAST/TESS/interm_tess_prf_retrieve/tess_prf_retrieval.ipynb]: #Convert -> # Convert

File: notebooks/MAST/TESS/interm_tess_prf_retrieve/tess_prf_retrieval.ipynb [hunk 16]
```diff
@@ -0,0 +1,852 @@
...
+    "    colOffset, rowOffset = getOffsetsFromPixelFractions(col, row)\n",
+    "    \n",
+    "    #Determine the 4 (col,row) locations with exact PRF measurements.\n",
+    "    imagePos = determineFourClosestPrfLoc(col, row)\n",
+    "\n",
+    "    #Loop over the 4 locations and read in each file and extract the sub-pixel location.\n",
```
Comment (ceb8) [inline notebooks/MAST/TESS/interm_tess_prf_retrieve/tess_prf_retrieval.ipynb]: #Loop -> # Loop

File: notebooks/MAST/TESS/interm_tess_prf_retrieve/tess_prf_retrieval.ipynb [hunk 17]
```diff
@@ -0,0 +1,852 @@
...
+    "    for pos in imagePos:\n",
+    "            prfArray = readOnePrfFitsFile(ccd, camera, pos[0], pos[1], path, datestring)\n",
+    "            \n",
+    "            img = getRegSampledPrfFitsByOffset(prfArray, colOffset, rowOffset)       \n",
+    "            prfImages.append(img)\n",
+    "\n",
+    "    #Simple linear interpolate across the 4 locations.\n",
```
Comment (ceb8) [inline notebooks/MAST/TESS/interm_tess_prf_retrieve/tess_prf_retrieval.ipynb]: #Simple -> # Simple

I'm going to stop pointing these out now, but all the comments that start with no space should have a space.

File: notebooks/MAST/TESS/interm_tess_prf_retrieve/tess_prf_retrieval.ipynb [hunk 18]
```diff
@@ -0,0 +1,852 @@
...
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "#Define the location for which we want to retrieve the PRF.\n",
+    "col = 125.0\n",
+    "row = 1044.0\n",
+    "ccd = 2\n",
+    "camera = 2\n",
+    "sector = 1\n",
+    "#This is the directory where MAST stores the prf FITS files.\n",
```
Comment (ceb8) [inline notebooks/MAST/TESS/interm_tess_prf_retrieve/tess_prf_retrieval.ipynb]: Suggest newline before this line for readability.

File: notebooks/MAST/TESS/interm_tess_prf_retrieve/tess_prf_retrieval.ipynb [hunk 19]
```diff
@@ -0,0 +1,852 @@
...
+    "path = \"https://archive.stsci.edu/missions/tess/models/prf_fitsfiles/\"\n",
+    "\n",
+    "kwargs['vmax'] = .2  \n",
+    "plt.figure(figsize = (14, 14))\n",
+    "#Loop over the 5 different locations\n",
```
Comment (ceb8) [inline notebooks/MAST/TESS/interm_tess_prf_retrieve/tess_prf_retrieval.ipynb]: Suggest newline before this line for readability.

File: notebooks/MAST/TESS/interm_tess_prf_retrieve/tess_prf_retrieval.ipynb [hunk 20]
```diff
@@ -0,0 +1,852 @@
...
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "#Define the location for which we want to retrieve the PRF.\n",
+    "col = 125.2\n",
+    "row = 544.1\n",
+    "ccd = 2\n",
+    "camera = 2\n",
+    "sector = 1\n",
+    "#This is the directory where MAST stores the prf FITS files.\n",
```
Comment (ceb8) [inline notebooks/MAST/TESS/interm_tess_prf_retrieve/tess_prf_retrieval.ipynb]: Suggest newline before this line for readability.

File: notebooks/MAST/TESS/interm_tess_prf_retrieve/tess_prf_retrieval.ipynb (hunk 1 above)
Comment (mustaric) [reply notebooks/MAST/TESS/interm_tess_prf_retrieve/tess_prf_retrieval.ipynb]: Added.

File: notebooks/MAST/TESS/interm_tess_prf_retrieve/tess_prf_retrieval.ipynb (hunk 2 above)
Comment (mustaric) [reply notebooks/MAST/TESS/interm_tess_prf_retrieve/tess_prf_retrieval.ipynb]: done.

File: notebooks/MAST/TESS/interm_tess_prf_retrieve/tess_prf_retrieval.ipynb (hunk 3 above)
Comment (mustaric) [reply notebooks/MAST/TESS/interm_tess_prf_retrieve/tess_prf_retrieval.ipynb]: sure.

File: notebooks/MAST/TESS/interm_tess_prf_retrieve/tess_prf_retrieval.ipynb (hunk 17 above)
Comment (mustaric) [reply notebooks/MAST/TESS/interm_tess_prf_retrieve/tess_prf_retrieval.ipynb]: Some day I will figure out why there is a standard one whether there are spaces or not.  Until then, I'll keep adding spaces all over the place.

**Comment by mustaric** [comment]
@ceb8  I think I have made all of your changes and added a requirements.txt file.

**Pull request merged**